corehq/apps/app_manager/suite_xml/sections/entries.py

```diff
@@ -879,24 +876,26 @@ def add_parent_datums(self, datums, module):
         datum_ids = {d.datum.id: d for d in datums}
         datums_by_case_tag = _get_datums_by_case_tag(datums)
 
-        for i, (this_datum_meta, parent_datum_meta) in list(enumerate(zip_longest(datums, parent_datums))):
-            if not parent_datum_meta or this_datum_meta == parent_datum_meta:
-                continue
-
-            if not parent_datum_meta.requires_selection:
-                # Add parent datums of opened subcases and automatically-selected cases
-                datums.insert(i, attr.evolve(parent_datum_meta, from_parent=True))
-            elif _same_case(this_datum_meta, parent_datum_meta) and this_datum_meta.action:
-                def set_id(datum, new_id):
-                    case_tag = getattr(this_datum_meta.action, 'case_tag', 'basic')
-                    _update_refs(datums_by_case_tag[case_tag], datum.id, new_id)
-                    datum.id = new_id
-
-                if parent_datum_meta.datum.id in datum_ids:
-                    datum = datum_ids[parent_datum_meta.datum.id]
-                    set_id(datum.datum, '_'.join((datum.datum.id, datum.case_type)))
-
-                set_id(this_datum_meta.datum, parent_datum_meta.datum.id)
+        ret = []
+        for this_datum_meta, parent_datum_meta in zip_longest(datums, parent_datums):
+            if parent_datum_meta and this_datum_meta != parent_datum_meta:
+                if not parent_datum_meta.requires_selection:
+                    # Add parent datums of opened subcases and automatically-selected cases
+                    ret.append(attr.evolve(parent_datum_meta, from_parent=True))
+                elif _same_case(this_datum_meta, parent_datum_meta) and this_datum_meta.action:
+                    def set_id(datum, new_id):
+                        case_tag = getattr(this_datum_meta.action, 'case_tag', 'basic')
+                        _update_refs(datums_by_case_tag[case_tag], datum.id, new_id)
+                        datum.id = new_id
+
+                    if parent_datum_meta.datum.id in datum_ids:
+                        datum = datum_ids[parent_datum_meta.datum.id]
+                        set_id(datum.datum, '_'.join((datum.datum.id, datum.case_type)))
+
+                    set_id(this_datum_meta.datum, parent_datum_meta.datum.id)
+            if this_datum_meta:
+                ret.append(this_datum_meta)
```

Comment:
I think this is incorrect as it will interleave the parent module datums with the child modules. I thought the 'datums' for the child module could all be added after the loop:

```
ret.extend(datums)
```

But that also breaks other tests. I think this whole method maybe needs to be re-structured to not 'zip' the datums (or we just leave it to mutate the list).